### PR TITLE
Use `iter_files` instead of `str(Path(...)` in image dataset

### PR DESCRIPTION
**Pull request by mariosasko**
Use `iter_files` in the `beans` and the `cats_vs_dogs` dataset scripts as suggested by @albertvillanova.

Additional changes:
* Fix `iter_files` in `MockDownloadManager` (see this [CI error](https://app.circleci.com/pipelines/github/huggingface/datasets/9247/workflows/2657ff8a-b531-4fd9-a9fc-6541a72e8d83/jobs/57028))
* Add support for `os.path.isdir` and `os.path.isfile` in streaming (`os.path.isfile` is needed in `StreamingDownloadManager`'s `iter_files` to make `cats_vs_dogs` streamable)

TODO:
- [ ] add tests for `xisdir` and `xisfile`

### CODE REVIEW

File: src/datasets/utils/streaming_download_manager.py [hunk 1]
```diff
@@ -160,6 +160,52 @@ def xbasename(a):
...
+        return os.path.isfile(path)
+    else:
+        if rest_hops and fsspec.get_fs_token_paths(rest_hops[0])[0].protocol == "https":
+            storage_options = {
+                "https": {"headers": get_authentication_headers_for_url(rest_hops[0], use_auth_token=use_auth_token)}
+            }
+        else:
+            storage_options = None
+        fs, *_ = fsspec.get_fs_token_paths(path, storage_options=storage_options)
+        return fs.isfile(path)
```
Comment (albertvillanova) [inline src/datasets/utils/streaming_download_manager.py:183]:
```suggestion
        return fs.isfile(main_hop)
```

File: src/datasets/utils/streaming_download_manager.py [hunk 2]
```diff
@@ -160,6 +160,52 @@ def xbasename(a):
...
+        return os.path.isdir(path)
+    else:
+        if rest_hops and fsspec.get_fs_token_paths(rest_hops[0])[0].protocol == "https":
+            storage_options = {
+                "https": {"headers": get_authentication_headers_for_url(rest_hops[0], use_auth_token=use_auth_token)}
+            }
+        else:
+            storage_options = None
+        fs, *_ = fsspec.get_fs_token_paths(path, storage_options=storage_options)
+        return fs.isdir(path)
```
Comment (albertvillanova) [inline src/datasets/utils/streaming_download_manager.py:206]:
```suggestion
        return fs.isdir(main_hop)
```

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks for all the enhancements, @mariosasko.

As discussed, the workaround in `iter_files` was temporal (it was valid only in some limited uses cases) and eventually would need fixing. In that moment I prioritized having a limited-working implementation rather than a general one. 

To have a generally valid implementation, first some fixing is needed in the custom FileSystems we have at `datasets`. I could have a look at that in another PR, so that you could merge it into this once that is fixed.

On the meantime, I have suggested some fixing to your implementation below.

**Review by albertvillanova — CHANGES_REQUESTED** [review]
This PR requires first being merged:
- #3481

**Review by albertvillanova — APPROVED** [review]
The failing test has nothing to do with this PR.

**Comment by lhoestq** [comment]
`iter_archive` is about to support ZIP archives. I think we should use this no ?

**Comment by lhoestq** [comment]
I'm not a big fan of having two functions that do the same thing. What do you think ?

**Comment by albertvillanova** [comment]
They do not do the same thing:
- One iterates over files in a directory
- The other I guess will iterate over the members of an archive

**Comment by lhoestq** [comment]
Makes sense ! Sounds good then - sorry for my misunderstanding

Note that `iter_archive` will be more performant for data streaming that `iter_files` thanks to the buffering so maybe in the future we can `iter_archive` for some of these datasets

**Comment by albertvillanova** [comment]
Yes, @lhoestq I agree with you: once `iter_archive` supports zip files, it will be more suitable than `iter_files` for these 2 datasets.

Anyway, this PR also implements `isfile`/`isdir` in streaming mode, besides fixing `iter_files`. And I'm interested in having those in master.

Maybe, could we merge this PR into master and take note to refactor the datasets to use `iter_archive` once zip is supported?
Other option could be to split this PR into 2...

File: src/datasets/streaming.py [hunk 3]
```diff
@@ -71,6 +73,9 @@ def wrapper(*args, **kwargs):
     patch_submodule(module, "os.path.join", xjoin).start()
     patch_submodule(module, "os.path.dirname", xdirname).start()
     patch_submodule(module, "os.path.basename", xbasename).start()
+    # allow checks on paths
+    patch_submodule(module, "os.path.isfile", wrap_auth(xisdir)).start()
```
Comment (lhoestq) [inline src/datasets/streaming.py:77]:
```suggestion
    patch_submodule(module, "os.path.isdir", wrap_auth(xisdir)).start()
```

**Review by lhoestq — COMMENTED** [review]
Sure, let me add some tests for xisdir and xisfile and we can merge

**Review by lhoestq — APPROVED** [review]
Alright let's merge this now :)

**Pull request merged**